core/tee/tee_svc.c

```diff
@@ -146,114 +150,348 @@ static const uint32_t fw_impl_bin_version; /* 0 by default */
 /* Trusted firmware manufacturer name */
 static const char fw_manufacturer[] = TO_STR(CFG_TEE_FW_MANUFACTURER);
 
+static TEE_Result get_prop_tee_dev_id(struct tee_ta_session *sess,
+				      void *buf, size_t blen)
+{
+	TEE_Result res;
+	TEE_UUID uuid;
+	const size_t nslen = 5;
+	uint8_t data[5 + FVR_DIE_ID_NUM_REGS * sizeof(uint32_t)] = {
+	    'O', 'P', 'T', 'E', 'E' };
+
+	if (blen < sizeof(uuid))
+		return TEE_ERROR_SHORT_BUFFER;
+
+	if (tee_otp_get_die_id(data + nslen, sizeof(data) - nslen))
+		return TEE_ERROR_BAD_STATE;
+
+	res = tee_hash_createdigest(TEE_ALG_SHA256, data, sizeof(data),
+				    (uint8_t *)&uuid, sizeof(uuid));
+	if (res != TEE_SUCCESS)
+		return TEE_ERROR_BAD_STATE;
+
+	/*
+	 * Changes the random value into and UUID as specifiec
+	 * in RFC 4122. The magic values are from the example
+	 * code in the RFC.
+	 *
+	 * TEE_UUID is defined slightly different from the RFC,
+	 * but close enough for our purpose.
+	 */
+
+	uuid.timeHiAndVersion &= 0x0fff;
+	uuid.timeHiAndVersion |= 5 << 12;
+
+	/* uuid.clock_seq_hi_and_reserved in the RFC */
+	uuid.clockSeqAndNode[0] &= 0x3f;
+	uuid.clockSeqAndNode[0] |= 0x80;
+
+	return tee_svc_copy_to_user(sess, buf, &uuid, sizeof(TEE_UUID));
+}
+
+static TEE_Result get_prop_tee_sys_time_prot_level(struct tee_ta_session *sess,
+						   void *buf, size_t blen)
+{
+	uint32_t prot;
+
+	if (blen < sizeof(prot))
+		return TEE_ERROR_SHORT_BUFFER;
+	prot = tee_time_get_sys_time_protection_level();
+	return tee_svc_copy_to_user(sess, (void *)buf, &prot, sizeof(prot));
+}
+
+static TEE_Result get_prop_client_id(struct tee_ta_session *sess,
+				     void *buf, size_t blen)
+{
+	if (blen < sizeof(TEE_Identity))
+		return TEE_ERROR_SHORT_BUFFER;
+	return tee_svc_copy_to_user(sess, buf, &sess->clnt_id,
+				    sizeof(TEE_Identity));
+}
+
+static TEE_Result get_prop_ta_app_id(struct tee_ta_session *sess,
+				     void *buf, size_t blen)
+{
+	if (blen < sizeof(TEE_UUID))
+		return TEE_ERROR_SHORT_BUFFER;
+	return tee_svc_copy_to_user(sess, buf, &sess->ctx->uuid,
+				    sizeof(TEE_UUID));
+}
+
 struct tee_props {
+	TEE_PropSetHandle prop_set;	/* part of TEE_PROPSET_XXX */
+	const char *name;
+
+	/* prop_type is of type enum user_ta_prop_type*/
+	uint32_t prop_type;
+
+	/* either get_prop_func or both data and len */
+	TEE_Result (*get_prop_func)(struct tee_ta_session *sess,
+				    void *buf, size_t blen);
 	const void *data;
 	const size_t len;
 };
 
 /* Consistent with enum utee_property */
 const struct tee_props tee_props_lut[] = {
```

Comment:
If we had one `struct tee_props` array per `prop_set` we could get rid of the `TEE_PropSetHandle prop_set` element in the struct and also use the `index` variable as a true index.